File: src/test/kotlin/dev/farneser/tasktracker/scheduler/config/TestContainerConfig.java
package dev.farneser.tasktracker.scheduler.config;

import com.github.dockerjava.api.model.ExposedPort;
import com.github.dockerjava.api.model.HostConfig;
import com.github.dockerjava.api.model.PortBinding;
import com.github.dockerjava.api.model.Ports;
import jakarta.annotation.PreDestroy;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.jdbc.datasource.DriverManagerDataSource;
import org.testcontainers.containers.PostgreSQLContainer;

import javax.sql.DataSource;

@Configuration
public class TestContainerConfig {
    @Bean
    public DataSource dataSource() {
        var pgContainer = new PostgreSQLContainer<>("postgres:latest")
                .withDatabaseName("task-tracker")
                .withUsername("postgres")
                .withPassword("postgres")
                .withExposedPorts(5432)
                .withCreateContainerCmdModifier(cmd -> cmd.withHostConfig(
                        new HostConfig().withPortBindings(new PortBinding(Ports.Binding.bindPort(15432), new ExposedPort(5432)))));

        pgContainer.start();

        var dataSource = new DriverManagerDataSource();

        dataSource.setDriverClassName("org.postgresql.Driver");
        dataSource.setUrl(pgContainer.getJdbcUrl());
        dataSource.setUsername(pgContainer.getUsername());
        dataSource.setPassword(pgContainer.getPassword());

        return dataSource;
    }
}

File: src/test/kotlin/dev/farneser/tasktracker/scheduler/config/TestContainerConfig.kt
package dev.farneser.tasktracker.scheduler.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.jdbc.datasource.DriverManagerDataSource
import org.testcontainers.containers.PostgreSQLContainer
import javax.sql.DataSource

@Configuration
class TestContainerConfig {
    @Bean
    fun dataSource(): DataSource {
        val pgContainer =
            PostgreSQLContainer("postgres:latest")
                .withDatabaseName("task-tracker")
                .withUsername("postgres")
                .withPassword("postgres")

        pgContainer.start()

        val dataSource = DriverManagerDataSource()

        dataSource.setDriverClassName("org.postgresql.Driver")
        dataSource.url = pgContainer.jdbcUrl
        dataSource.username = pgContainer.username
        dataSource.password = pgContainer.password

        return dataSource
    }
}

